File: deepeval/cli/auth/flow.py
"""Browser authentication and CLI-native onboarding for `deepeval login`.

Device-code style pairing (RFC 8628 shaped), with no localhost callback
server. The CLI never handles credentials and no secrets ever transit
localhost, so the flow works over SSH / containers / Colab and is unaffected
by Chrome's Local Network Access permission prompt.

Flow:

1. The CLI asks the backend to create a short-lived login session:

       POST {api_base}/cli/auth/sessions
       -> 200 {"success": true,
               "data": {"userCode": "ABCD-1234",
                        "deviceCode": "<opaque high-entropy token>",
                        "expiresIn": 600,
                        "interval": 3}}

   `userCode` is short and user-visible (shown in the terminal and in the
   browser so the user can confirm they're pairing the right session).
   `deviceCode` is a secret with >= 128 bits of entropy; it never leaves the
   CLI and is the only thing that can claim the API key.

2. The CLI opens the browser at the returned verification URL. The browser
   handles authentication and explicit authorization only; onboarding remains
   in the terminal.

3. The CLI polls the token endpoint with the secret device code:

       POST {api_base}/cli/auth/sessions/token
            {"deviceCode": "<token>"}
       -> pending:   {"success": true, "data": {"status": "pending"}}
       -> authenticated: {"success": true,
                      "data": {"status": "authenticated",
                               "setupToken": "<short-lived token>",
                               "email": "user@example.com"}}
       -> expired / consumed / unknown: 404, or
                     {"success": false, "error": "..."}

4. The CLI fetches onboarding context and completes setup using the short-lived
   setup token. The completion response contains a newly minted project API
   key, and consumes the login session.
"""

import sys
import time
import webbrowser
from typing import Any, Dict, List, Optional, Sequence, Tuple, Union

import requests
import typer
from pydantic import ValidationError
from rich import print

import deepeval
from deepeval.cli.auth.api import (
    CliAuthorization,
    CliOnboardingContext,
    DynamicNewUserOnboardingRequest,
    ExistingProjectKeyRequest,
    DevicePairing,
)
from deepeval.cli.utils import with_utm
from deepeval.confident.api import get_base_api_url
from deepeval.telemetry import set_logged_in_with

CREATE_PAIRING_ENDPOINT = "/cli/auth/sessions"
PAIRING_TOKEN_ENDPOINT = "/cli/auth/sessions/token"
ONBOARDING_ENDPOINT = "/cli/onboarding"
ONBOARDING_COMPLETE_ENDPOINT = "/cli/onboarding/complete"

REQUEST_TIMEOUT_SECONDS = 10
# Completion creates the organization/project in one backend transaction,
# which can take longer than a normal request. Retries are safe because the
# request carries an idempotency key.
COMPLETE_TIMEOUT_SECONDS = 30
COMPLETE_MAX_ATTEMPTS = 3


class AuthFlowError(Exception):
    """Raised when the browser login flow cannot complete."""


def _is_interactive() -> bool:
    try:
        return sys.stdin.isatty() and sys.stdout.isatty()
    except Exception:
        return False


CONFIDENT_PURPLE = "#6a00ff"


def _fill_radio_indicator(question: Any) -> None:
    """Make the select behave like a radio group: ● on the pointed row, ○ on
    the rest.

    questionary's `use_indicator` renders a hollow circle on every row in
    select mode (the filled one is reserved for checkbox mode, driven by
    `selected_options`, which select never populates). Sync `selected_options`
    with the pointed row at render time so the active option gets the filled
    dot.
    """
    from questionary.prompts.common import InquirerControl

    for control in question.application.layout.find_all_controls():
        if isinstance(control, InquirerControl):
            original = control.text

            def patched(control=control, original=original):
                control.selected_options = [control.get_pointed_at().value]
                return original()

            control.text = patched
            return


def _questionary_style():
    import questionary

    return questionary.Style(
        [
            ("qmark", f"fg:{CONFIDENT_PURPLE} bold"),
            ("question", "bold"),
            ("instruction", "fg:#767676"),
            ("pointer", f"fg:{CONFIDENT_PURPLE} bold"),
            ("highlighted", f"fg:{CONFIDENT_PURPLE} bold"),
            # prompt_toolkit's built-in style renders "selected" in
            # reverse video (fg/bg swapped); noreverse disables that.
            ("selected", f"fg:{CONFIDENT_PURPLE} bold noreverse"),
            ("answer", f"fg:{CONFIDENT_PURPLE} bold"),
        ]
    )


def prompt_text(message: str, default: Optional[str] = None) -> str:
    """Styled text input matching `prompt_select`, with a plain fallback."""
    if _is_interactive():
        try:
            import questionary

            answer = questionary.text(
                message,
                default=default or "",
                qmark="?",
                style=_questionary_style(),
            ).ask()
            # `ask()` returns None when the user aborts (e.g. Ctrl-C).
            if answer is None:
                raise AuthFlowError("Input cancelled.")
            return answer
        except AuthFlowError:
            raise
        except Exception:
            # Any questionary/terminal issue: degrade to the plain prompt.
            pass

    return typer.prompt(message, default=default)


def prompt_select(message: str, choices: Sequence[Tuple[str, Any]]) -> Any:
    """Arrow-key selection when the terminal supports it, numbered fallback
    otherwise."""
    if _is_interactive():
        try:
            import questionary

            style = _questionary_style()
            question = questionary.select(
                message,
                choices=[
                    questionary.Choice(title=label, value=value)
                    for label, value in choices
                ],
                qmark="?",
                pointer="❯",
                use_indicator=True,
                instruction="(arrow keys)",
                style=style,
            )
            _fill_radio_indicator(question)
            answer = question.ask()
            # `ask()` returns None when the user aborts (e.g. Ctrl-C).
            if answer is None:
                raise AuthFlowError("Selection cancelled.")
            return answer
        except AuthFlowError:
            raise
        except Exception:
            # Any questionary/terminal issue: degrade to the numbered prompt.
            pass

    print(message)
    for index, (label, _) in enumerate(choices, start=1):
        print(f"  {index}. {label}")
    while True:
        raw = typer.prompt("Enter a number")
        try:
            selected = int(raw)
        except (TypeError, ValueError):
            print("❌ Please enter a valid number.")
            continue
        if 1 <= selected <= len(choices):
            return choices[selected - 1][1]
        print(f"❌ Please enter a number between 1 and {len(choices)}.")


def prompt_checkbox(
    message: str,
    choices: Sequence[Tuple[str, Any]],
    min_selections: int = 1,
) -> List[Any]:
    """Arrow-key multi-selection with a comma-separated numbered fallback."""
    if _is_interactive():
        try:
            import questionary

            answer = questionary.checkbox(
                message,
                choices=[
                    questionary.Choice(title=label, value=value)
                    for label, value in choices
                ],
                qmark="?",
                pointer="❯",
                instruction="(space to select, enter to continue)",
                style=_questionary_style(),
                validate=lambda values: len(values) >= min_selections
                or f"Select at least {min_selections} option(s).",
            ).ask()
            if answer is None:
                raise AuthFlowError("Selection cancelled.")
            return list(answer)
        except AuthFlowError:
            raise
        except Exception:
            # Any questionary/terminal issue: degrade to the numbered prompt.
            pass

    print(message)
    for index, (label, _) in enumerate(choices, start=1):
        print(f"  {index}. {label}")
    while True:
        if min_selections == 0:
            raw = typer.prompt(
                "Enter numbers (comma-separated), or leave blank",
                default="",
                show_default=False,
            )
            if not raw.strip():
                return []
        else:
            raw = typer.prompt("Enter one or more numbers (comma-separated)")
        try:
            selected = [int(value.strip()) for value in raw.split(",")]
        except (AttributeError, TypeError, ValueError):
            print("❌ Enter valid numbers separated by commas.")
            continue
        if (
            len(selected) >= min_selections
            and len(selected) == len(set(selected))
            and all(1 <= value <= len(choices) for value in selected)
        ):
            return [choices[value - 1][1] for value in selected]
        print(
            f"❌ Select at least {min_selections} unique number(s) between "
            f"1 and {len(choices)}, separated by commas."
        )


def _request_headers() -> Dict[str, str]:
    return {
        "Content-Type": "application/json",
        "X-DeepEval-Version": deepeval.__version__,
    }


def _unwrap(res: requests.Response) -> Dict[str, Any]:
    """Unwrap the standard `{success, data, error}` API envelope. Tolerates a
    bare payload so the backend can be iterated on without breaking the CLI."""
    try:
        payload = res.json()
    except ValueError:
        raise AuthFlowError(
            f"{res.request.method} {res.url} did not return JSON: "
            f"{res.text[:200]!r}"
        )
    if not isinstance(payload, dict):
        raise AuthFlowError(
            f"{res.request.method} {res.url} returned an unexpected payload: "
            f"{res.text[:200]!r}"
        )
    if "success" in payload:
        if not payload.get("success"):
            raise AuthFlowError(
                payload.get("error") or "The pairing request failed."
            )
        data = payload.get("data")
        return data if isinstance(data, dict) else {}
    return payload


def create_pairing() -> DevicePairing:
    url = f"{get_base_api_url()}{CREATE_PAIRING_ENDPOINT}"
    try:
        res = requests.post(
            url,
            headers=_request_headers(),
            json={},
            timeout=REQUEST_TIMEOUT_SECONDS,
        )
    except requests.RequestException as e:
        raise AuthFlowError(f"Could not reach POST {url} ({e}).")

    if res.status_code == 404:
        raise AuthFlowError(
            f"The backend does not support browser login yet "
            f"(404 from POST {url})."
        )
    if res.status_code != 200:
        raise AuthFlowError(
            f"POST {url} failed with HTTP {res.status_code}: {res.text[:200]!r}"
        )

    data = _unwrap(res)
    try:
        return DevicePairing.model_validate(data)
    except ValidationError:
        raise AuthFlowError(
            f"POST {url} did not return the required login session fields: "
            f"{res.text[:200]!r}"
        )


def _poll_once(
    device_code: str,
) -> Optional[CliAuthorization]:
    """One poll of the token endpoint.

    Returns setup authorization when browser auth completes, None while pending
    (or on a transient network error), and raises AuthFlowError on terminal
    failures (expired, consumed, backend missing).
    """
    url = f"{get_base_api_url()}{PAIRING_TOKEN_ENDPOINT}"
    try:
        res = requests.post(
            url,
            headers=_request_headers(),
            json={"deviceCode": device_code},
            timeout=REQUEST_TIMEOUT_SECONDS,
        )
    except requests.RequestException:
        return None

    if res.status_code == 404:
        raise AuthFlowError(
            f"This pairing is no longer valid, or the backend does not "
            f"support browser login yet (404 from POST {url})."
        )
    if res.status_code != 200:
        raise AuthFlowError(
            f"POST {url} failed with HTTP {res.status_code}: {res.text[:200]!r}"
        )

    data = _unwrap(res)
    status = data.get("status")
    if status == "pending":
        return None
    if status == "authenticated":
        try:
            return CliAuthorization.model_validate(data)
        except ValidationError:
            raise AuthFlowError(
                "Browser authentication completed but the server did not "
                "return a setup token."
            )
    raise AuthFlowError(
        f"The pairing is no longer valid (status: {status or 'unknown'}). "
        "Run `deepeval login` again."
    )


def browser_pairing_login() -> Optional[CliAuthorization]:
    """Authenticate the user in the browser and return CLI setup access.

    Returns a short-lived setup authorization, or None if the flow was aborted
    / timed out / unavailable.
    """
    try:
        pairing = create_pairing()
    except AuthFlowError:
        print(
            "\n⚠️  Unexpected error — seems like browser login isn't "
            "available right now."
        )
        return None

    login_url = with_utm(
        pairing.verification_url,
        medium="cli",
        content="login_pair_browser_open",
    )
    fallback_url = with_utm(
        pairing.verification_url,
        medium="cli",
        content="login_pair_fallback_link",
    )

    print("\n🌐 Opening your browser — confirm this pairing code there:")
    print(f"\n    [bold green]{pairing.user_code}[/bold green]\n")
    webbrowser.open(login_url)
    print(
        f"[dim]Browser didn't open? Use [link={fallback_url}]{fallback_url}[/link][/dim]"
    )
    print(
        "[dim]Waiting for the browser... "
        "press Ctrl+C to paste an API key instead.[/dim]"
    )

    deadline = time.monotonic() + pairing.expires_in
    try:
        while time.monotonic() < deadline:
            time.sleep(pairing.interval)
            try:
                completed = _poll_once(pairing.device_code)
            except AuthFlowError as e:
                print(f"\n⚠️  {e}")
                return None
            if completed is None:
                continue
            if completed.email:
                set_logged_in_with(completed.email)
            return completed
        print("\n⌛ Timed out waiting for the browser.")
        return None
    except KeyboardInterrupt:
        print()
        return None


def get_cli_onboarding_context(setup_token: str) -> CliOnboardingContext:
    url = f"{get_base_api_url()}{ONBOARDING_ENDPOINT}"
    try:
        res = requests.get(
            url,
            headers={
                **_request_headers(),
                "Authorization": f"Bearer {setup_token}",
            },
            timeout=REQUEST_TIMEOUT_SECONDS,
        )
    except requests.RequestException as e:
        raise AuthFlowError(f"Could not reach GET {url} ({e}).")
    if res.status_code != 200:
        raise AuthFlowError(
            f"GET {url} failed with HTTP {res.status_code}: {res.text[:200]!r}"
        )
    try:
        data = res.json()
    except ValueError:
        raise AuthFlowError(f"GET {url} did not return JSON.")
    try:
        return CliOnboardingContext.model_validate(data)
    except ValidationError:
        raise AuthFlowError(
            f"GET {url} returned an unexpected onboarding payload."
        )


def complete_cli_onboarding(
    setup_token: str,
    request: Union[
        DynamicNewUserOnboardingRequest,
        ExistingProjectKeyRequest,
    ],
    idempotency_key: str,
) -> str:
    url = f"{get_base_api_url()}{ONBOARDING_COMPLETE_ENDPOINT}"
    res = None
    last_error: Optional[requests.RequestException] = None
    for attempt in range(COMPLETE_MAX_ATTEMPTS):
        retryable = attempt < COMPLETE_MAX_ATTEMPTS - 1
        try:
            res = requests.post(
                url,
                headers={
                    **_request_headers(),
                    "Authorization": f"Bearer {setup_token}",
                    "Idempotency-Key": idempotency_key,
                },
                json=request.to_payload(),
                timeout=COMPLETE_TIMEOUT_SECONDS,
            )
        except requests.RequestException as e:
            last_error = e
            if retryable:
                print("Still working — retrying...")
                time.sleep(2)
            continue
        # A previous (timed-out) attempt may still hold the completion lock
        # server-side; give it a moment and retry with the same key.
        if res.status_code == 409 and "in progress" in res.text and retryable:
            time.sleep(3)
            continue
        break
    if res is None:
        raise AuthFlowError(f"Could not reach POST {url} ({last_error}).")
    if res.status_code != 200:
        raise AuthFlowError(
            f"POST {url} failed with HTTP {res.status_code}: "
            f"{res.text[:200]!r}"
        )
    try:
        data = res.json()
    except ValueError:
        raise AuthFlowError(f"POST {url} did not return JSON.")
    if not isinstance(data, dict) or not data.get("apiKey"):
        raise AuthFlowError(
            "CLI onboarding completed without returning an API key."
        )
    return str(data["apiKey"])

File: typescript/src/cli/auth/flow.ts
// Browser authentication and CLI onboarding for `deepeval login`, ported from
// deepeval/cli/auth/flow.py. Device-code pairing (RFC 8628 shaped) with no
// localhost callback server, so the flow works over SSH and in containers.

import axios from "axios";
import { getBaseApiUrl } from "@/confident/api";
import { getVersion } from "@/cli/version";
import { openBrowser, withUtm } from "@/cli/utm";
import { prompt } from "@/cli/utils";

const CREATE_PAIRING_ENDPOINT = "/cli/auth/sessions";
const PAIRING_TOKEN_ENDPOINT = "/cli/auth/sessions/token";
const ONBOARDING_ENDPOINT = "/cli/onboarding";
const ONBOARDING_COMPLETE_ENDPOINT = "/cli/onboarding/complete";

const REQUEST_TIMEOUT_MS = 10_000;
// Completion creates the org and project in one backend transaction, which can
// outlast a normal request.
const COMPLETE_TIMEOUT_MS = 30_000;
const COMPLETE_MAX_ATTEMPTS = 3;

const DEFAULT_EXPIRES_IN_SECONDS = 600;
const DEFAULT_POLL_INTERVAL_SECONDS = 3;

export class AuthFlowError extends Error {}

export interface DevicePairing {
  userCode: string;
  deviceCode: string;
  verificationUrl: string;
  expiresIn: number;
  interval: number;
}

export interface CliAuthorization {
  setupToken: string;
  email?: string;
}

export interface CliOnboardingProject {
  id: string;
  name: string;
  canCreateApiKey: boolean;
}

export type QuestionnaireAnswer = string | boolean | string[];

export interface CliQuestionnaireOption {
  label: string;
  value: string | boolean;
  exclusive?: boolean;
  acceptsCustomValue?: boolean;
  customPrompt?: string;
}

export interface CliQuestionnaireQuestion {
  id: string;
  prompt: string;
  required: boolean;
  type: "text" | "single_select" | "multi_select";
  options?: CliQuestionnaireOption[];
  defaultValue?: string | boolean;
  maxLength?: number;
  minSelections?: number;
}

export interface CliQuestionnaire {
  version: number;
  questions: CliQuestionnaireQuestion[];
}

export interface CliOnboardingContext {
  state: "new_user" | "existing_user";
  projects: CliOnboardingProject[];
  questionnaire?: CliQuestionnaire;
}

function requestHeaders(): Record<string, string> {
  return {
    "Content-Type": "application/json",
    "X-DeepEval-Version": getVersion(),
  };
}

/** Unwraps the `{success, data, error}` envelope, tolerating a bare payload. */
function unwrap(data: unknown, what: string): Record<string, unknown> {
  if (!data || typeof data !== "object") {
    throw new AuthFlowError(`${what} returned an unexpected payload.`);
  }
  const payload = data as Record<string, unknown>;
  if ("success" in payload) {
    if (!payload.success) {
      throw new AuthFlowError(
        (payload.error as string) || "The pairing request failed.",
      );
    }
    const inner = payload.data;
    return inner && typeof inner === "object"
      ? (inner as Record<string, unknown>)
      : {};
  }
  return payload;
}

async function createPairing(): Promise<DevicePairing> {
  const url = `${getBaseApiUrl()}${CREATE_PAIRING_ENDPOINT}`;
  let response;
  try {
    response = await axios.post(
      url,
      {},
      { headers: requestHeaders(), timeout: REQUEST_TIMEOUT_MS },
    );
  } catch (error) {
    const status = (error as { response?: { status?: number } }).response
      ?.status;
    if (status === 404) {
      throw new AuthFlowError(
        `The backend does not support browser login yet (404 from POST ${url}).`,
      );
    }
    throw new AuthFlowError(`Could not reach POST ${url}.`);
  }

  const data = unwrap(response.data, `POST ${url}`);
  const userCode = data.userCode as string | undefined;
  const deviceCode = data.deviceCode as string | undefined;
  const verificationUrl = data.verificationUriComplete as string | undefined;
  if (!userCode || !deviceCode || !verificationUrl) {
    throw new AuthFlowError(
      `POST ${url} did not return the required login session fields.`,
    );
  }
  return {
    userCode,
    deviceCode,
    verificationUrl,
    expiresIn: (data.expiresIn as number) ?? DEFAULT_EXPIRES_IN_SECONDS,
    interval: (data.interval as number) ?? DEFAULT_POLL_INTERVAL_SECONDS,
  };
}

/** `null` while pending or on a transient error; throws on terminal failures. */
async function pollOnce(deviceCode: string): Promise<CliAuthorization | null> {
  const url = `${getBaseApiUrl()}${PAIRING_TOKEN_ENDPOINT}`;
  let response;
  try {
    response = await axios.post(
      url,
      { deviceCode },
      { headers: requestHeaders(), timeout: REQUEST_TIMEOUT_MS },
    );
  } catch (error) {
    const status = (error as { response?: { status?: number } }).response
      ?.status;
    if (status === 404) {
      throw new AuthFlowError(
        "This pairing is no longer valid, or the backend does not support " +
          "browser login yet.",
      );
    }
    return null;
  }

  const data = unwrap(response.data, `POST ${url}`);
  const status = data.status as string | undefined;
  if (status === "pending") return null;
  if (status === "authenticated") {
    const setupToken = data.setupToken as string | undefined;
    if (!setupToken) {
      throw new AuthFlowError(
        "Browser authentication completed but the server did not return a " +
          "setup token.",
      );
    }
    return { setupToken, email: data.email as string | undefined };
  }
  throw new AuthFlowError(
    `The pairing is no longer valid (status: ${status ?? "unknown"}). ` +
      "Run `deepeval login` again.",
  );
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

/** Returns `null` when the flow was aborted, timed out, or is unavailable. */
export async function browserPairingLogin(): Promise<CliAuthorization | null> {
  let pairing: DevicePairing;
  try {
    pairing = await createPairing();
  } catch {
    console.log(
      "\n⚠️  Unexpected error — seems like browser login isn't available right now.",
    );
    return null;
  }

  const loginUrl = withUtm(pairing.verificationUrl, {
    content: "login_pair_browser_open",
  });
  const fallbackUrl = withUtm(pairing.verificationUrl, {
    content: "login_pair_fallback_link",
  });

  console.log("\n🌐 Opening your browser — confirm this pairing code there:");
  console.log(`\n    ${pairing.userCode}\n`);
  await openBrowser(loginUrl);
  console.log(`Browser didn't open? Use ${fallbackUrl}`);
  console.log("Waiting for the browser... press Ctrl+C to stop.");

  const deadline = Date.now() + pairing.expiresIn * 1000;
  while (Date.now() < deadline) {
    await sleep(pairing.interval * 1000);
    let completed: CliAuthorization | null;
    try {
      completed = await pollOnce(pairing.deviceCode);
    } catch (error) {
      console.log(`\n⚠️  ${(error as Error).message}`);
      return null;
    }
    if (completed) return completed;
  }
  console.log("\n⌛ Timed out waiting for the browser.");
  return null;
}

export async function getCliOnboardingContext(
  setupToken: string,
): Promise<CliOnboardingContext> {
  const url = `${getBaseApiUrl()}${ONBOARDING_ENDPOINT}`;
  let response;
  try {
    response = await axios.get(url, {
      headers: { ...requestHeaders(), Authorization: `Bearer ${setupToken}` },
      timeout: REQUEST_TIMEOUT_MS,
    });
  } catch (error) {
    throw new AuthFlowError(
      `Could not reach GET ${url} (${(error as Error).message}).`,
    );
  }
  const data = response.data as Partial<CliOnboardingContext> | undefined;
  if (!data || (data.state !== "new_user" && data.state !== "existing_user")) {
    throw new AuthFlowError(`GET ${url} returned an unexpected payload.`);
  }
  return {
    state: data.state,
    projects: data.projects ?? [],
    questionnaire: data.questionnaire,
  };
}

export type OnboardingRequest =
  | {
      questionnaireVersion: number;
      questionnaireAnswers: Record<string, QuestionnaireAnswer>;
    }
  | { projectId: string };

export async function completeCliOnboarding(
  setupToken: string,
  request: OnboardingRequest,
  idempotencyKey: string,
): Promise<string> {
  const url = `${getBaseApiUrl()}${ONBOARDING_COMPLETE_ENDPOINT}`;
  let lastError: unknown = null;

  for (let attempt = 0; attempt < COMPLETE_MAX_ATTEMPTS; attempt++) {
    const retryable = attempt < COMPLETE_MAX_ATTEMPTS - 1;
    try {
      const response = await axios.post(url, request, {
        headers: {
          ...requestHeaders(),
          Authorization: `Bearer ${setupToken}`,
          "Idempotency-Key": idempotencyKey,
        },
        timeout: COMPLETE_TIMEOUT_MS,
      });
      const apiKey = (response.data as { apiKey?: string } | undefined)?.apiKey;
      if (!apiKey) {
        throw new AuthFlowError(
          "CLI onboarding completed without returning an API key.",
        );
      }
      return apiKey;
    } catch (error) {
      if (error instanceof AuthFlowError) throw error;
      lastError = error;
      const response = (
        error as { response?: { status?: number; data?: unknown } }
      ).response;
      // A timed-out attempt may still hold the completion lock server-side;
      // the idempotency key makes retrying safe.
      const inProgress =
        response?.status === 409 &&
        JSON.stringify(response.data ?? "").includes("in progress");
      if (retryable) {
        if (!inProgress) console.log("Still working — retrying...");
        await sleep(inProgress ? 3000 : 2000);
        continue;
      }
    }
  }

  throw new AuthFlowError(
    `Could not complete POST ${url} (${
      lastError instanceof Error ? lastError.message : "unknown error"
    }).`,
  );
}

// Numbered prompts, the fallback questionary degrades to in Python when the
// terminal isn't interactive.

export async function promptText(
  message: string,
  defaultValue?: string,
): Promise<string> {
  const suffix = defaultValue ? ` [${defaultValue}]` : "";
  const answer = await prompt(`? ${message}${suffix} `);
  return answer || defaultValue || "";
}

export async function promptSelect<T>(
  message: string,
  choices: Array<[string, T]>,
): Promise<T> {
  console.log(message);
  choices.forEach(([label], index) => console.log(`  ${index + 1}. ${label}`));
  for (;;) {
    const raw = await prompt("Enter a number: ");
    const selected = Number(raw);
    if (
      Number.isInteger(selected) &&
      selected >= 1 &&
      selected <= choices.length
    ) {
      return choices[selected - 1][1];
    }
    console.log(`❌ Please enter a number between 1 and ${choices.length}.`);
  }
}

export async function promptCheckbox<T>(
  message: string,
  choices: Array<[string, T]>,
  minSelections = 1,
): Promise<T[]> {
  console.log(message);
  choices.forEach(([label], index) => console.log(`  ${index + 1}. ${label}`));
  for (;;) {
    const raw = await prompt(
      minSelections === 0
        ? "Enter numbers (comma-separated), or leave blank: "
        : "Enter one or more numbers (comma-separated): ",
    );
    if (!raw.trim() && minSelections === 0) return [];
    const parsed = raw
      .split(",")
      .map((value) => Number(value.trim()))
      .filter((value) => Number.isInteger(value));
    const unique = new Set(parsed);
    const valid =
      parsed.length >= minSelections &&
      unique.size === parsed.length &&
      parsed.every((value) => value >= 1 && value <= choices.length);
    if (valid) return parsed.map((value) => choices[value - 1][1]);
    console.log(
      `❌ Select at least ${minSelections} unique number(s) between 1 and ` +
        `${choices.length}, separated by commas.`,
    );
  }
}
